File: skyspark-scala/src/main/scala/com/github/dkanellis/skyspark/scala/api/algorithms/bnl/BnlAlgorithm.scala
package com.github.dkanellis.skyspark.scala.api.algorithms.bnl

import com.github.dkanellis.skyspark.scala.api.algorithms.{DominatingAlgorithm, Point}

import scala.collection.mutable.ListBuffer

private[bnl] class BnlAlgorithm extends Serializable {

  private[bnl] def computeSkylinesWithPreComparison(flagsWithPoints: Iterable[(Flag, Point)]): Iterable[Point] = {
    val localSkylines = ListBuffer[Point]()
    flagsWithPoints
      .filter(fp => passesPreComparison(fp._1))
      .map(_._2)
      .foreach(addDiscardOrDominate(localSkylines, _))

    localSkylines
  }

  private def passesPreComparison(flag: Flag) = {
    var passes = true
    for (i <- 0 until flag.size) {
      passes &= flag.bit(i)
    }

    !passes
  }

  private[bnl] def computeSkylinesWithoutPreComparison(pointIterable: Iterable[Point]): Iterable[Point] = {
    val localSkylines = ListBuffer[Point]()
    for (candidateSkyline <- pointIterable) {
      addDiscardOrDominate(localSkylines, candidateSkyline)
    }

    localSkylines
  }

  private def addDiscardOrDominate(localSkylines: ListBuffer[Point], candidateSkyline: Point) {
    for (pointToCheckAgainst <- localSkylines) {
      if (DominatingAlgorithm.dominates(pointToCheckAgainst, candidateSkyline)) {
        return
      } else if (DominatingAlgorithm.dominates(candidateSkyline, pointToCheckAgainst)) {
        localSkylines -= pointToCheckAgainst
      }
    }

    localSkylines += candidateSkyline
  }
}

File: skyspark-java/src/main/java/com/akikanellis/skyspark/api/algorithms/bnl/BnlAlgorithm.java
package com.akikanellis.skyspark.api.algorithms.bnl;

import com.akikanellis.skyspark.api.algorithms.Point;
import com.google.common.annotations.VisibleForTesting;
import scala.Tuple2;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;

@VisibleForTesting
public class BnlAlgorithm implements Serializable {

    Iterable<Point> computeSkylinesWithoutPreComparison(Iterable<Point> points) {
        List<Point> skylines = new ArrayList<>();

        points.forEach(p -> addDiscardOrDominate(skylines, p));

        return skylines;
    }

    private void addDiscardOrDominate(List<Point> currentSkylines, Point candidateSkyline) {
        Iterator<Point> currentSkylinesIterator = currentSkylines.iterator();
        while (currentSkylinesIterator.hasNext()){
            Point pointToCheckAgainst = currentSkylinesIterator.next();

            if (pointToCheckAgainst.dominates(candidateSkyline)) return;

            if (candidateSkyline.dominates(pointToCheckAgainst)) currentSkylinesIterator.remove();
        }

        currentSkylines.add(candidateSkyline);
    }

    Iterable<Point> computeSkylinesWithPreComparison(Collection<Tuple2<Flag, Point>> flagsPoints) {
        List<Point> skylines = new ArrayList<>();

        flagsPoints.stream()
                .filter(fp -> passesPreComparison(fp._1()))
                .map(Tuple2::_2)
                .forEach(p -> addDiscardOrDominate(skylines, p));

        return skylines;
    }

    private boolean passesPreComparison(Flag flag) {
        for (int i = 0; i < flag.size(); i++) {
            if (!flag.bit(i)) return true;
        }

        return false;
    }
}
